waspc/data/Generator/templates/sdk/wasp/client/operations/rpc.ts

```diff
@@ -0,0 +1,38 @@
+import { type Route } from "wasp/client";
+import type {
+  _Awaited,
+  _ReturnType,
+} from "wasp/universal/types"
+
+// PRIVATE API (but should maybe be public, users define values of this type)
+export type Query<Input, Output> = QueryFunction<Input, Output> & QueryMetadata
+
+// PRIVATE API (but should maybe be public, users define values of this type)
+export type Action<Input, Output> = ClientOperation<Input, Output>
+
+// PRIVATE API
+export type QueryFunction<Input, Output> = ClientOperation<Input, Output>
+
+// PRIVATE API
+export type QueryMetadata = {
+  queryCacheKey: string[]
+  route: Route
+}
+
+// PRIVATE API (needed in SDK)
+export type OperationRpcFor<BackendOperation extends GenericBackendOperation> =
+  Parameters<BackendOperation> extends []
```

Comment:
A [playground link](https://www.typescriptlang.org/play/?noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&ts=4.9.5#code/PTAEBUAsEsGdTqAhqW0C2AHANtAZtAKYAmoAboQE5oD2AdqDXqAO6RIAurhoxNCDGgFdKoAMY1ihADSgARkI4AoEKkjDspQgA8cSaAw6QelQklj15hGHVJGeAAwAKSSknSEOVWAB4AQkhiANaEtgDymFSc0PQAfKA6XrbwANoAug4qYBK20BwxdAB0SlmgAILkNNCkEVH5llIEdCSseZCgDrDRsASEsA4JyaBCmMjwRogcAJ6RpQAUmJQ0kaJkVaR4QnRi9XQAlErTkaAAausAYltioAC8oHN7t-F0QuhyVCWqUIiTxqCLNDE2H0HlIsBGVCOPCYyGw2FAtTcu1gxR0mBolC4UNAAHFQlRoGIAsFQsREdFLHc5q4AOawABcoGaFEoshyXm0HEZSDoU0eN3iWyCdBoLDonzAAFEyEhsEJOH1QBx+AByDiUISEFVWMRIISwHjgADKStMnHgDjO1Uu2wG5lQQjk2KYpQceOalEJxJC4RWFLoAwAtKh+HhXCHQDSaJJitircQbWJJZzSbBvaTybtbkpQKcLlcEimhu6CUTAj6yX6swB+U2a0CMsPYA0S0AAdXoatYGKCOr1Brz1oL7HgIv+rncnm8sntaMIOxIsZmPBcbnQ8Duq8nXmoPihMPjidiKQADGlW35FAhmCweJgr-Zx2up6JEm4dgUBMrkJH8Z7rlCpRGJwSpICE8AoNitKvKEXA5LA6r6HQXh2Pw7zIKU4JOsujDMG6f5euWGZVgUDhzBM8DsJgkR0PAbChNwKqmPIQjQJoBg0owIilAASk4ADCSo4Y0Bh5AUsB7IUUmHDheIcOc0DUBwW7oD45yFkkxDwCW-7pr6dQFPENw5qAKkvr45zHmerZSYU7Y8IEYjVLBsrYFMpqEr2xpfvwEx0EEHEdIeVydE+24fKouAIQgsB0F2zRiH0XSem5CRYNMxQgKUCl0K5UwzgwhDpe5rhuO5UxEJo8BbCJzTELIbCEu0BgSJQphAtMoDAu8zZKn8mBIFM2A0EgdjLqUB75tsYwdDVhBNCQDhLscLhDUgkhhHgwXTXcckKUpKk+NtYixK2UBUPNGIyNwTKEC0KA0vKlB2OwcHGMEdHGPYoiPgNz47l1cBcDCKCbNsuylIgRWYBlMnHHtikISpABMakaamuIEWWJL6UihnZrmZk7hZ8SJBj6SgLWazVA2pkTh4xNqVZ56lGUth9T88DU-VrAwGIzWfSBt6sDyKHLSug3AhtW1TWIyO3Jj8mI8p9Oo8dp2szSSENTwfAhh4GluKgHBCHgzDvGgUjwG+gT5HQnGPgQSlhQzVDSKUl5YjA8C-DwrU8M5KDvDYL2OET3j+ERuP+qTRZaaA6SZKoOTEGJ9DFEoQA) that explains why we need `Parameters<BackendOperation> extends []`.